### PR TITLE
Dev

### LINKED ISSUE #207: Integrate Mitiq into OpenQAOA
### Prerequisites
Before raising this feature request, I have made sure:

- [ ] It does not lead to duplication.
- [ ] It has not already been requested.

### PR DESCRIPTION
**Pull request by victor-onofre**
## Description

  - Integrate Mitiq into OpenQAOA to allow error mitigation techniques to take place during the execution of the QAOA. We have created a wrapper focusing on the ZNE technique following a similar structure to the SPAM Twirling wrapper. 
- This is a project born from the Quantum Open Source Foundation Mentorship program 
- New dependency to add: Mitiq
- **Fixes #207**

## Checklist

[//]: <> (- [ ] My code follows the style guidelines of this project)

- [x] I have performed a self-review of my code.
- [x] I have commented my code and used numpy-style docstrings
- [ ] I have made corresponding updates to the documentation.
- [x] My changes generate no new warnings
- [ ] I have added/updated tests to make sure bugfix/feature works.
- [x] New and existing unit tests pass locally with my changes.

[//]: <> (- [ ] Any dependent changes have been merged and published in downstream modules)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

We have been testing the wrapper with the Max Cut problem but we need help on defining unit-tests. What will be the best test to add?

### CODE REVIEW

**Comment by victor-onofre** [comment]
Hi! As I mentioned in the description, we need some help defining the unit tests for the integration. At the moment, we are running multiple tests with the Max Cut problem using a noise model, and different parameters of the ZNE

File: src/openqaoa-core/openqaoa/algorithms/workflow_properties.py
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/algorithms/workflow_properties.py:294]: Add the `str` that triggers creation of a `Spam_Twirling_ErrorMitigationProperties` object.

File: src/openqaoa-core/openqaoa/algorithms/workflow_properties.py
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/algorithms/workflow_properties.py:353]: Add the `str` that triggers creation of a `Mitiq_Zne_ErrorMitigationProperties` object.

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/backends/wrapper.py:13]: Is this used in the code?

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/backends/wrapper.py:8]: Is this used in the code?

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/backends/wrapper.py:118]: It would be nice to raise a `ValueError` rather than an `AssertionError` in these cases, what do you think?

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/backends/wrapper.py:129]: We are doing case testing here, we can raise the error message here instead if the `factory`, `scaling` or any other parameter is not valid

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/backends/wrapper.py:163]: Typo, too many r's

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/backends/wrapper.py:111]: Same comment here, maybe raise a `ValueError` instead of an `AssertionError`

**Review by KilianPoirier — COMMENTED** [review]
Overall the PR looks pretty good and I don't think fundamental changes are required here. Most comments concern docstrings and cosmetic changes to the code you implemented.

Feel free to justify your choices if you feel like my comments wouldn't improve the quality/readability of the code.

Two things are still missing from the PR:

- Passing the current tests. I need you to merge the `dev` branch back into yours so the tests that were fixed by the latest update pass.
- Adding unit tests as discussed.

We can also think about updating the [website's page dedicated to error mitigation](https://openqaoa.entropicalabs.com/error-mitigation/spam-twirling/) with explanations on ZNE for potential users.
I was thinking it would be nice to have an example notebook using either SPAM or ZNE or an example for both. Do you think this is possible too? ([see other example notebooks for reference](https://github.com/entropicalabs/openqaoa/tree/main/examples))

File: src/openqaoa-core/openqaoa/algorithms/baseworkflow.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/baseworkflow.py]: Added in the last commits

File: src/openqaoa-braket/openqaoa_braket/backends/qaoa_braket_qpu.py
Comment (mar-ven) [reply src/openqaoa-braket/openqaoa_braket/backends/qaoa_braket_qpu.py:210]: Yes, solved in the last commits

File: src/openqaoa-core/openqaoa/optimizers/training_vqa.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/optimizers/training_vqa.py]: Yes, solved in the last commits

File: src/openqaoa-pyquil/openqaoa_pyquil/backends/qaoa_pyquil_qpu.py
Comment (mar-ven) [reply src/openqaoa-pyquil/openqaoa_pyquil/backends/qaoa_pyquil_qpu.py]: Yes, solved in the last commits

File: src/openqaoa-qiskit/openqaoa_qiskit/backends/qaoa_qiskit_qpu.py
Comment (mar-ven) [reply src/openqaoa-qiskit/openqaoa_qiskit/backends/qaoa_qiskit_qpu.py:231]: Yes, solved in the last commits

File: src/openqaoa-core/openqaoa/algorithms/baseworkflow.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/baseworkflow.py]: Addressed in the last commits

File: src/openqaoa-core/openqaoa/algorithms/baseworkflow.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/baseworkflow.py:310]: Removed line `self.error_mitigation_properties = Mitiq_Zne_ErrorMitigationProperties()`
and also line `self.error_mitigation_properties = Spam_Twirling_ErrorMitigationProperties()`

The following for loop only acts on the temporary variable error_mitigation_properties, and in case it succeeds, the corresponding object is created and set into the Workflow internal attribute.

File: src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_result.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/qaoa/qaoa_result.py:314]: Solved in the last commits

File: src/openqaoa-core/openqaoa/algorithms/workflow_properties.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/workflow_properties.py]: Addressed in the last commits

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (victor-onofre) [inline src/openqaoa-core/openqaoa/backends/wrapper.py:114]: I have noticed that we cannot use the statevector simulator with our current implementation. We used the get_counts to obtain the results which you cannot use with the statevector

File: src/openqaoa-core/openqaoa/algorithms/workflow_properties.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/workflow_properties.py:269]: Addressed in the last commits

File: src/openqaoa-core/openqaoa/algorithms/workflow_properties.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/workflow_properties.py:284]: Check is already performed into set_error_mitigation_properties from Workflow class (openqaoa-core/openqaoa/algorithms/baseworkflow.py, line 299)

File: src/openqaoa-core/openqaoa/algorithms/workflow_properties.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/workflow_properties.py:287]: Addressed in the last commits, also renamed Mitiq ZNE Properties class to MitiqZNEProperties to be more concise

File: src/openqaoa-core/openqaoa/algorithms/workflow_properties.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/workflow_properties.py:294]: I don't understand what modifications to perform. Please verify if the last commits addressed this issue as well

File: src/openqaoa-core/openqaoa/algorithms/workflow_properties.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/workflow_properties.py:353]: I don't understand what modifications to perform. Please verify if the last commits addressed this issue as well

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/backends/wrapper.py:13]: Addressed in the last commits

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/backends/wrapper.py:119]: Solved in the last commits

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/backends/wrapper.py:130]: We already raise a ValueError previously, if parameters are not valid. I think that raising a further exception here would be redundant. What do you think?

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/backends/wrapper.py:164]: Addressed in the last commits

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/backends/wrapper.py:112]: Addressed in the last commits

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/backends/wrapper.py:114]: Addressed in the last commits

File: src/openqaoa-core/openqaoa/algorithms/baseworkflow.py
Comment (mar-ven) [reply src/openqaoa-core/openqaoa/algorithms/baseworkflow.py]: I had to add the following fix: `self.error_mitigation_properties = error_mitigation_properties()` after the if-else block. This is necessary: python can only check if an attribute belongs to an instance, not to a class. We necessarily have to create an empty instance and then verify if the attribute belongs to that instance.

**Comment by victor-onofre** [comment]
Hi @KilianPoirier,

All the comments have been addressed, and the `dev` branch has been merged. We have added two tests similar to the SPAM wrapper, it is still necessary to add a couple more, and we plan to do it next week. I think you can now review all these changes. 

The example notebook is a good idea. Do we do that in this PR or create another one? 

Thanks!

File: src/openqaoa-core/openqaoa/algorithms/workflow_properties.py
Comment (KilianPoirier) [reply src/openqaoa-core/openqaoa/algorithms/workflow_properties.py:353]: I meant describing which input `error_mitigation_technique` takes here, but that may be redundant.

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (KilianPoirier) [reply src/openqaoa-core/openqaoa/backends/wrapper.py:130]: What I mean is that the check is redundant in the sense that you first check if your factory is in the available factories line:118 and then you match the string to the supported factories. What you could do is raise a `ValueError` in the last `else` statement.
The same comment can probably be applied to `scale_noise`.

**Review by KilianPoirier — COMMENTED** [review]
A few more comments here.
I approved the test pipeline run, so you'll have more info if the test are passing or not now that you have merged dev back into your branches.

File: src/openqaoa-core/openqaoa/algorithms/baseworkflow.py
Comment (KilianPoirier) [reply src/openqaoa-core/openqaoa/algorithms/baseworkflow.py]: Alright that makes sense, let's leave it as is

**Comment by victor-onofre** [comment]
Hi @KilianPoirier, 

I have added `mitiq` to the requirements, but the tests are still waiting for approval. Some of the previous errors were related to that issue.  

We are going to work on the example notebooks and the rest of the tests. 

Thanks!

**Comment by KilianPoirier** [comment]
Hi @victor-onofre ,

Thanks for updating the requirements.
Good news, the tests are passing! Let me know when you implement your own tests and the example notebook.

**Comment by victor-onofre** [comment]
Hi @KilianPoirier! We have push an update with the example notebook and the tests. Now is ready for another review. 

Thanks!

**Comment by KilianPoirier** [comment]
Hey @victor-onofre , it seems the regular tests are failing now. 

It is most likely due to the use of images in your example notebook. Can you try copying the images directly in the notebook without using PIL? 
I believe that if the notebook is ran correctly and the image is resolved, the image itself will be tracked in the notebook without a need to import it.

I will also have a look at the docs and let you know if I find anything.

File: src/openqaoa-core/openqaoa/backends/wrapper.py
Comment (KilianPoirier) [inline src/openqaoa-core/openqaoa/backends/wrapper.py:114]: Are you saying that `'qiskit.statevector_simulator'` is not supported?

typo for exception

File: src/openqaoa-qiskit/tests/test_qiskit_backend_wrapper.py
Comment (KilianPoirier) [inline src/openqaoa-qiskit/tests/test_qiskit_backend_wrapper.py:396]: If Qiskit is the only backend supported, does it make sense to ignore an import error here?

**Review by KilianPoirier — COMMENTED** [review]
Only minor comments. Good job on the implementation of tests and the recent changes!

I have a clue where the issue lies for the docs tests, try to add the following line in the toctree `openqaoa/docs/source/index.rst`, l.306
```
   notebooks/15_Zero_Noise_Extrapolation.ipynb
```

**Review by KilianPoirier — APPROVED** [review]
Looks good to me!

**Pull request merged**
